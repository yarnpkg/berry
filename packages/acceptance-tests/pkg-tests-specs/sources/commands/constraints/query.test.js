const {
  fs: {writeFile},
} = require(`pkg-tests-core`);

const {environments} = require(`./environments`);

const queries = {
  [`single predicate`]: `dependency_type(DependencyType).`,
  [`single predicate with ignored variable`]: `workspace_ident(_, WorkspaceName).`,
  [`combined predicates`]: `dependency_type(DependencyType), workspace_has_dependency(_, DependencIdent, DependencyRange, DependencyType).`,
  [`custom predicate`]: `custom_predicate(DependencyType).`,
  [`filter w/ workspace_field_test/3`]: `workspace(WorkspaceCwd), workspace_field_test(WorkspaceCwd, 'name', '$$ === "workspace-a"').`,
  [`filter w/ workspace_field_test/4`]: `workspace(WorkspaceCwd), workspace_field_test(WorkspaceCwd, 'name', '$$ === $0', ['workspace-b']).`,
};

const constraintsFile = `
custom_predicate(DependencyType):-
  dependency_type(DependencyType),
  DependencyType \\= dependencies.
`;

describe(`Commands`, () => {
  describe(`constraints query`, () => {
    test(
      `test without trailing .`,
      makeTemporaryEnv({}, {
        plugins: [
          require.resolve(`@yarnpkg/monorepo/scripts/plugin-constraints.js`),
        ],
      }, async({path, run, source}) => {
        await environments[`one regular dependency`](path);

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`constraints`, `query`, `workspace_ident(_, WorkspaceName)`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      }),
    );

    test(
      `test with a syntax error`,
      makeTemporaryEnv({}, {
        plugins: [
          require.resolve(`@yarnpkg/monorepo/scripts/plugin-constraints.js`),
        ],
      }, async({path, run, source}) => {
        await environments[`one regular dependency`](path);

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`constraints`, `query`, `*&%@$#$#!$@)`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      }),
    );

    test(
      `test with an unknown predicate`,
      makeTemporaryEnv({}, {
        plugins: [
          require.resolve(`@yarnpkg/monorepo/scripts/plugin-constraints.js`),
        ],
      }, async({path, run, source}) => {
        await environments[`one regular dependency`](path);

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`constraints`, `query`, `hello_word(X)`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      }),
    );

    test(
      `test with an empty predicate`,
      makeTemporaryEnv({}, {
        plugins: [
          require.resolve(`@yarnpkg/monorepo/scripts/plugin-constraints.js`),
        ],
      }, async({path, run, source}) => {
        await environments[`one regular dependency`](path);

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`constraints`, `query`, `workspace()`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      }),
    );

    for (const [environmentDescription, environment] of Object.entries(environments)) {
      for (const [queryDescription, query] of Object.entries(queries)) {
        test(
          `test (${environmentDescription} / ${queryDescription})`,
          makeTemporaryEnv({}, {
            plugins: [
              require.resolve(`@yarnpkg/monorepo/scripts/plugin-constraints.js`),
            ],
          }, async ({path, run, source}) => {
            await environment(path);
            await writeFile(`${path}/constraints.pro`, constraintsFile);

            let code;
            let stdout;
            let stderr;

            try {
              ({code, stdout, stderr} = await run(`constraints`, `query`, query));
            } catch (error) {
              ({code, stdout, stderr} = error);
            }

            expect({code, stdout, stderr}).toMatchSnapshot();
          }),
        );
      }
    }
  });
});
