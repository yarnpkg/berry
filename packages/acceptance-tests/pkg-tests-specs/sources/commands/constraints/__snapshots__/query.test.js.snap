// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands constraints query test (empty project / combined predicates) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (empty project / custom predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (empty project / single predicate with ignored variable) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: WorkspaceName = 'root-workspace-988eec'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (empty project / single predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = dependencies
➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (multiple workspaces / combined predicates) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (multiple workspaces / custom predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (multiple workspaces / single predicate with ignored variable) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: WorkspaceName = 'root-workspace-988eec'
➤ BR0000: WorkspaceName = 'workspace-a-01beb5'
➤ BR0000: WorkspaceName = 'workspace-b-d95f30'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (multiple workspaces / single predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = dependencies
➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (one regular dependency / combined predicates) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (one regular dependency / custom predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (one regular dependency / single predicate with ignored variable) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: WorkspaceName = 'root-workspace-988eec'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (one regular dependency / single predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = dependencies
➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two development dependencies / combined predicates) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two development dependencies / custom predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two development dependencies / single predicate with ignored variable) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: WorkspaceName = 'root-workspace-988eec'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two development dependencies / single predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = dependencies
➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies / combined predicates) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies / custom predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies / single predicate with ignored variable) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: WorkspaceName = 'root-workspace-988eec'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies / single predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = dependencies
➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies, two development dependencies / combined predicates) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = dependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: ┌ DependencyType  = devDependencies
➤ BR0000: │ DependencIdent  = 'no-deps-bin'
➤ BR0000: └ DependencyRange = '1.0.0'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies, two development dependencies / custom predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies, two development dependencies / single predicate with ignored variable) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: WorkspaceName = 'root-workspace-988eec'
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test (two regular dependencies, two development dependencies / single predicate) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: DependencyType = dependencies
➤ BR0000: DependencyType = devDependencies
➤ BR0000: DependencyType = peerDependencies
➤ BR0000: Done
",
}
`;

exports[`Commands constraints query test without trailing . 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ BR0000: Done
",
}
`;
